## Observations

<b>Positive</b>
- Data types of all the columns seems correct except "Sent_date" which is string
- All the flag columns values are either 1 or 0
- is_sent column value is always 1
- 


<b>Negative</b>
- Had issue while reading the csv file. Encoding other than utf-8 passed
- All the columns except "Primary Topic" are not not null. (Find the reason why primary topic is NULL)
- Understand more about email campign id as it is not PK.
- "Primary Topic" Column has special characters that doesn't mean anything (/, --). Spelling errors

In [1]:
import pandas as pd

In [2]:
email_path = "../data/Email_Campaign_Details.csv"
# email_df = pd.read_csv(email_path, encoding='cp1252')
email_df = pd.read_csv(email_path, encoding='mac_roman')
email_df.head()

Email_Campaign_id          contact_id               email_subject  \
0  701200000006cIGAAY  00v2000000IZVYWAA5  Roadshow Houston (Advisor)   
1  701200000006cIGAAY  00v2000000IZW13AAH  Roadshow Houston (Advisor)   
2  701200000006cIGAAY  00v2000000IZUepAAH  Roadshow Houston (Advisor)   
3  701200000006cIGAAY  00v2000000IZWDdAAP  Roadshow Houston (Advisor)   
4  701200000006cIGAAY  00v2000000IZUdSAAX  Roadshow Houston (Advisor)   

   is_sent  is_delivered  is_link_clicked  is_hard_bounce  is_bounce  \
0        1             1                0               0          0   
1        1             1                0               0          0   
2        1             1                0               0          0   
3        1             1                0               0          0   
4        1             1                0               0          0   

   is_opt_out  is_open Sent_date       Campaign_type Primary Topic  
0           0        1  2/7/2023  Events - In Person             /  
1           0        1  2/7/2023  Events - In Person             /  
2           0        0  2/7/2023  Events - In Person             /  
3           0        0  2/7/2023  Events - In Person             /  
4           0        0  2/7/2023  Events - In Person             /

In [3]:
email_df.dtypes

Email_Campaign_id    object
contact_id           object
email_subject        object
is_sent               int64
is_delivered          int64
is_link_clicked       int64
is_hard_bounce        int64
is_bounce             int64
is_opt_out            int64
is_open               int64
Sent_date            object
Campaign_type        object
Primary Topic        object
dtype: object

In [4]:
email_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128236 entries, 0 to 128235
Data columns (total 13 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   Email_Campaign_id  128236 non-null  object
 1   contact_id         128236 non-null  object
 2   email_subject      128236 non-null  object
 3   is_sent            128236 non-null  int64 
 4   is_delivered       128236 non-null  int64 
 5   is_link_clicked    128236 non-null  int64 
 6   is_hard_bounce     128236 non-null  int64 
 7   is_bounce          128236 non-null  int64 
 8   is_opt_out         128236 non-null  int64 
 9   is_open            128236 non-null  int64 
 10  Sent_date          128236 non-null  object
 11  Campaign_type      128236 non-null  object
 12  Primary Topic      111816 non-null  object
dtypes: int64(7), object(6)
memory usage: 12.7+ MB


In [5]:
email_df.describe(include = 'all')

Email_Campaign_id          contact_id email_subject   is_sent  \
count               128236              128236        128236  128236.0   
unique                 514                3872           466       NaN   
top     70120000000MfxSAAS  00v2000000IZVYWAA5  ETF Review 8       NaN   
freq                  3778                  35          3778       NaN   
mean                   NaN                 NaN           NaN       1.0   
std                    NaN                 NaN           NaN       0.0   
min                    NaN                 NaN           NaN       1.0   
25%                    NaN                 NaN           NaN       1.0   
50%                    NaN                 NaN           NaN       1.0   
75%                    NaN                 NaN           NaN       1.0   
max                    NaN                 NaN           NaN       1.0   

         is_delivered  is_link_clicked  is_hard_bounce      is_bounce  \
count   128236.000000    128236.000000   128236.000000  128236.000000   
unique            NaN              NaN             NaN            NaN   
top               NaN              NaN             NaN            NaN   
freq              NaN              NaN             NaN            NaN   
mean         0.978243         0.004616        0.000686       0.022240   
std          0.145889         0.067788        0.026187       0.147465   
min          0.000000         0.000000        0.000000       0.000000   
25%          1.000000         0.000000        0.000000       0.000000   
50%          1.000000         0.000000        0.000000       0.000000   
75%          1.000000         0.000000        0.000000       0.000000   
max          1.000000         1.000000        1.000000       1.000000   

           is_opt_out        is_open Sent_date Campaign_type Primary Topic  
count   128236.000000  128236.000000    128236        128236        111816  
unique            NaN            NaN       219             6            18  
top               NaN            NaN  6/7/2023    Newsletter           ETF  
freq              NaN            NaN      6102         39060         34706  
mean         0.000094       0.260036       NaN           NaN           NaN  
std          0.009673       0.438656       NaN           NaN           NaN  
min          0.000000       0.000000       NaN           NaN           NaN  
25%          0.000000       0.000000       NaN           NaN           NaN  
50%          0.000000       0.000000       NaN           NaN           NaN  
75%          0.000000       1.000000       NaN           NaN           NaN  
max          1.000000       1.000000       NaN           NaN           NaN

There are only 514 unique email campaign ids. Need to understand more about email campaign id



### Duplicate Analysis

- The data seems to have duplicates for Campaign id : "70120000000MfxSAAS"
- Campaign id : "70120000000MfxSAAS" had 2 sent emails
- Upon verification in next section it doesn't seem to be duplicates.

In [6]:
email_df[email_df.duplicated()]

Email_Campaign_id          contact_id email_subject  is_sent  \
29968  70120000000MfxSAAS  00v2000000bnkNcAAI  ETF Review 8        1   
29969  70120000000MfxSAAS  00v2000000bGVGpAAO  ETF Review 8        1   
29970  70120000000MfxSAAS  00v2000000HPhJtAAL  ETF Review 8        1   
29971  70120000000MfxSAAS  00v2000000RD8MHAA1  ETF Review 8        1   
29972  70120000000MfxSAAS  00v2000000U2zWUAAZ  ETF Review 8        1   
...                   ...                 ...           ...      ...   
30315  70120000000MfxSAAS  00v2000000MinpkAAB  ETF Review 8        1   
30316  70120000000MfxSAAS  00v2000000MjeypAAB  ETF Review 8        1   
30317  70120000000MfxSAAS  00v2000000NawFcAAJ  ETF Review 8        1   
30318  70120000000MfxSAAS  00v2000000NavbNAAR  ETF Review 8        1   
30319  70120000000MfxSAAS  00v2000000OSMa3AAH  ETF Review 8        1   

       is_delivered  is_link_clicked  is_hard_bounce  is_bounce  is_opt_out  \
29968             1                0               0          0           0   
29969             1                0               0          0           0   
29970             1                0               0          0           0   
29971             1                0               0          0           0   
29972             1                0               0          0           0   
...             ...              ...             ...        ...         ...   
30315             1                0               0          0           0   
30316             1                0               0          0           0   
30317             1                0               0          0           0   
30318             1                0               0          0           0   
30319             1                0               0          0           0   

       is_open Sent_date Campaign_type Primary Topic  
29968        0  8/1/2023    Newsletter           ETF  
29969        0  8/1/2023    Newsletter           ETF  
29970        0  8/1/2023    Newsletter           ETF  
29971        0  8/1/2023    Newsletter           ETF  
29972        0  8/1/2023    Newsletter           ETF  
...        ...       ...           ...           ...  
30315        0  8/1/2023    Newsletter           ETF  
30316        1  8/1/2023    Newsletter           ETF  
30317        0  8/1/2023    Newsletter           ETF  
30318        0  8/1/2023    Newsletter           ETF  
30319        0  8/1/2023    Newsletter           ETF  

[208 rows x 13 columns]

In [7]:
email_df[(email_df["Email_Campaign_id"] == "70120000000MfxSAAS") \
& (email_df["contact_id"] == "00v2000000bnkNcAAI")]

Email_Campaign_id          contact_id email_subject  is_sent  \
26544  70120000000MfxSAAS  00v2000000bnkNcAAI  ETF Review 8        1   
29968  70120000000MfxSAAS  00v2000000bnkNcAAI  ETF Review 8        1   

       is_delivered  is_link_clicked  is_hard_bounce  is_bounce  is_opt_out  \
26544             1                0               0          0           0   
29968             1                0               0          0           0   

       is_open Sent_date Campaign_type Primary Topic  
26544        0  8/1/2023    Newsletter           ETF  
29968        0  8/1/2023    Newsletter           ETF

In [8]:
email_df[email_df.duplicated()][["Email_Campaign_id"]].value_counts()

Email_Campaign_id 
70120000000MfxSAAS    208
Name: count, dtype: int64

In [9]:
email_df[email_df.duplicated(keep=False)][["contact_id"]].value_counts()

contact_id        
00v2000000FD02sAAD    2
00v2000000FOMVUAA5    2
00v2000000ZAataAAD    2
00v2000000ZAbzZAAT    2
00v2000000ZAptfAAD    2
                     ..
00v2000000OSNjCAAX    2
00v2000000OSNwaAAH    2
00v2000000OSddiAAD    2
00v2000000OSf1HAAT    2
00vD000000h81NvIAI    2
Name: count, Length: 208, dtype: int64

### Email campaign - contact id relation

- Email_campaign_id and contact_id has one to one for all campaigns except "70120000000MfxSAAS"
  

In [10]:
email_df[email_df["Email_Campaign_id"] != "70120000000MfxSAAS"] \
[["Email_Campaign_id", "contact_id"]].value_counts()

Email_Campaign_id   contact_id        
701200000006JxXAAU  00v2000000IbcyJAAR    1
701D0000000NBOaIAO  00v2000000bnkNcAAI    1
                    00v2000000dj5k7AAA    1
                    00v2000000di4trAAA    1
                    00v2000000dgRm6AAE    1
                                         ..
70120000000Mb8YAAS  00vD000000h1PUKIA2    1
                    00vD000000h1PK9IAM    1
                    00vD000000h1MHwIAM    1
                    00vD000000h1HCkIAM    1
701D0000000NfA6IAK  00v2000000KNyzJAAT    1
Name: count, Length: 124458, dtype: int64

In [11]:
email_df[email_df["Email_Campaign_id"] == "70120000000MfxSAAS"] \
.sort_values(by = ["contact_id"])

Email_Campaign_id          contact_id email_subject  is_sent  \
29789  70120000000MfxSAAS  00v2000000FCqPzAAL  ETF Review 8        1   
30066  70120000000MfxSAAS  00v2000000FCzgJAAT  ETF Review 8        1   
26642  70120000000MfxSAAS  00v2000000FCzgJAAT  ETF Review 8        1   
26565  70120000000MfxSAAS  00v2000000FD02sAAD  ETF Review 8        1   
29989  70120000000MfxSAAS  00v2000000FD02sAAD  ETF Review 8        1   
...                   ...                 ...           ...      ...   
30204  70120000000MfxSAAS  00vD000000h81NvIAI  ETF Review 8        1   
26780  70120000000MfxSAAS  00vD000000h81NvIAI  ETF Review 8        1   
29795  70120000000MfxSAAS  00vD000000h81tOIAQ  ETF Review 8        1   
27812  70120000000MfxSAAS  00vD000000h82ZCIAY  ETF Review 8        1   
29692  70120000000MfxSAAS  00vD000000h85T1IAI  ETF Review 8        1   

       is_delivered  is_link_clicked  is_hard_bounce  is_bounce  is_opt_out  \
29789             0                0               0          1           0   
30066             1                0               0          0           0   
26642             1                0               0          0           0   
26565             1                0               0          0           0   
29989             1                0               0          0           0   
...             ...              ...             ...        ...         ...   
30204             1                0               0          0           0   
26780             1                0               0          0           0   
29795             1                0               0          0           0   
27812             1                0               0          0           0   
29692             1                0               0          0           0   

       is_open Sent_date Campaign_type Primary Topic  
29789        0  8/1/2023    Newsletter           ETF  
30066        1  8/1/2023    Newsletter           ETF  
26642        0  8/1/2023    Newsletter           ETF  
26565        1  8/1/2023    Newsletter           ETF  
29989        1  8/1/2023    Newsletter           ETF  
...        ...       ...           ...           ...  
30204        0  8/1/2023    Newsletter           ETF  
26780        0  8/1/2023    Newsletter           ETF  
29795        0  8/1/2023    Newsletter           ETF  
27812        0  8/1/2023    Newsletter           ETF  
29692        1  8/1/2023    Newsletter           ETF  

[3778 rows x 13 columns]

In [12]:
email_df[email_df["Email_Campaign_id"] == "70120000000MfxSAAS"]["contact_id"].value_counts()

contact_id
00v2000000bFmbLAAS    2
00v2000000U02sOAAR    2
00v2000000XQaqHAAT    2
00v2000000fl8QNAAY    2
00v2000000fGlQUAA0    2
                     ..
00v2000000KLfqtAAD    1
00v2000000KLjbkAAD    1
00v2000000KLuhRAAT    1
00v2000000KLmYEAA1    1
00v2000000feSGyAAM    1
Name: count, Length: 3424, dtype: int64

In [13]:
email_dup_rem = email_df.drop_duplicates()

In [14]:
email_dup_rem[["Email_Campaign_id", "contact_id"]].value_counts()

Email_Campaign_id   contact_id        
70120000000MfxSAAS  00v2000000WNg0kAAD    2
                    00v2000000QEbQeAAL    2
                    00v2000000TLOrkAAH    2
                    00v2000000aF7VHAA0    2
                    00v2000000aFCPIAA4    2
                                         ..
70120000000MdBcAAK  00v2000000dVswGAAS    1
                    00v2000000dUefUAAS    1
                    00v2000000d0pZRAAY    1
                    00v2000000cqUmzAAE    1
701D0000000NfA6IAK  00v2000000KNyzJAAT    1
Name: count, Length: 127882, dtype: int64

In [15]:
campaign_contact_df = pd.DataFrame(email_dup_rem[["Email_Campaign_id", "contact_id"]].value_counts())


In [16]:
campaign_contact_df[campaign_contact_df["count"] == 2]

count
Email_Campaign_id  contact_id               
70120000000MfxSAAS 00v2000000WNg0kAAD      2
                   00v2000000QEbQeAAL      2
                   00v2000000TLOrkAAH      2
                   00v2000000aF7VHAA0      2
                   00v2000000aFCPIAA4      2
...                                      ...
                   00v2000000NdBgBAAV      2
                   00v2000000OSLO4AAP      2
                   00vD000000h1HCkIAM      2
                   00v2000000Wm6NlAAJ      2
                   00v2000000Nd9ThAAJ      2

[146 rows x 1 columns]

In [17]:
email_df[(email_df["Email_Campaign_id"] == "70120000000MfxSAAS") \
& (email_df["contact_id"] == "00v2000000WNg0kAAD")]

Email_Campaign_id          contact_id email_subject  is_sent  \
26880  70120000000MfxSAAS  00v2000000WNg0kAAD  ETF Review 8        1   
30304  70120000000MfxSAAS  00v2000000WNg0kAAD  ETF Review 8        1   

       is_delivered  is_link_clicked  is_hard_bounce  is_bounce  is_opt_out  \
26880             1                0               0          0           0   
30304             1                0               0          0           0   

       is_open Sent_date Campaign_type Primary Topic  
26880        0  8/1/2023    Newsletter           ETF  
30304        1  8/1/2023    Newsletter           ETF

### Unique value for columns

In [18]:
email_df["Campaign_type"].value_counts()

Campaign_type
Newsletter                 39060
Marketing Campaign         32026
Event follow up            22497
Digital event              16538
Events - In Person         10024
Marketing Communication     8091
Name: count, dtype: int64

In [19]:
email_df["Primary Topic"].value_counts()

Primary Topic
ETF                34706
Bonds              14321
Portfolio          13840
Market Insights    13226
Ad hoc              8632
Adhoc               7167
Equities            6514
Insights            5326
HealthCare          4943
Asssets             1095
Funds                814
Income               465
Investment           449
Retirement           187
Asset                 61
--                    39
/                     25
Growth                 6
Name: count, dtype: int64

In [20]:
email_df.is_sent.value_counts()

is_sent
1    128236
Name: count, dtype: int64

In [21]:
email_df.is_delivered.value_counts()

is_delivered
1    125446
0      2790
Name: count, dtype: int64

### Special characters in Primary Tpoic Column

- Use the email subject to fill the primary topic to a relevant one
- / can be treated as NULL for "Primary Topic"
-  -- can be converted to NULL as other others are null as well

In [22]:
email_df[email_df["Primary Topic"] == "/"]

Email_Campaign_id          contact_id                 email_subject  \
0   701200000006cIGAAY  00v2000000IZVYWAA5    Roadshow Houston (Advisor)   
1   701200000006cIGAAY  00v2000000IZW13AAH    Roadshow Houston (Advisor)   
2   701200000006cIGAAY  00v2000000IZUepAAH    Roadshow Houston (Advisor)   
3   701200000006cIGAAY  00v2000000IZWDdAAP    Roadshow Houston (Advisor)   
4   701200000006cIGAAY  00v2000000IZUdSAAX    Roadshow Houston (Advisor)   
5   701200000006cIGAAY  00v2000000IZUkRAAX    Roadshow Houston (Advisor)   
6   701200000006cIGAAY  00v2000000IZVBLAA5    Roadshow Houston (Advisor)   
7   701200000006cIGAAY  00v2000000IZW96AAH    Roadshow Houston (Advisor)   
8   701200000006cIGAAY  00v2000000IZVVVAA5    Roadshow Houston (Advisor)   
9   701200000006cIGAAY  00v2000000IZWDXAA5    Roadshow Houston (Advisor)   
10  701200000006cIGAAY  00v2000000IZVgmAAH    Roadshow Houston (Advisor)   
11  701200000006cIGAAY  00v2000000IZW8gAAH    Roadshow Houston (Advisor)   
12  701200000006cIGAAY  00v2000000IZUjiAAH    Roadshow Houston (Advisor)   
13  701200000006cIGAAY  00v2000000IZW2PAAX    Roadshow Houston (Advisor)   
14  701200000006cIGAAY  00v2000000IZW04AAH    Roadshow Houston (Advisor)   
27  701200000006JxYAAU  00v2000000LNA9XAAX  Brief outlook on Real Estate   
28  701200000006JxYAAU  00v2000000MhuxqAAB  Brief outlook on Real Estate   
29  701200000006JxYAAU  00v2000000NbHl6AAF  Brief outlook on Real Estate   
30  701200000006JxYAAU  00v2000000P77GRAAZ  Brief outlook on Real Estate   
31  701200000006JxYAAU  00v2000000QGvTdAAL  Brief outlook on Real Estate   
32  701200000006JxYAAU  00v2000000SK6qeAAD  Brief outlook on Real Estate   
33  701200000006JxYAAU  00v2000000fgHbJAAU  Brief outlook on Real Estate   
34  701200000006JxYAAU  00v2000000PL3NlAAL  Brief outlook on Real Estate   
35  701200000006JxYAAU  00v2000000FkmBYAAZ  Brief outlook on Real Estate   
36  701200000006JxYAAU  00v2000000MiYl6AAF  Brief outlook on Real Estate   

    is_sent  is_delivered  is_link_clicked  is_hard_bounce  is_bounce  \
0         1             1                0               0          0   
1         1             1                0               0          0   
2         1             1                0               0          0   
3         1             1                0               0          0   
4         1             1                0               0          0   
5         1             1                0               0          0   
6         1             1                0               0          0   
7         1             1                0               0          0   
8         1             1                0               0          0   
9         1             1                0               0          0   
10        1             1                0               0          0   
11        1             1                0               0          0   
12        1             1                0               0          0   
13        1             1                0               0          0   
14        1             1                0               0          0   
27        1             1                0               0          0   
28        1             1                0               0          0   
29        1             1                0               0          0   
30        1             1                0               0          0   
31        1             1                0               0          0   
32        1             1                0               0          0   
33        1             1                0               0          0   
34        1             1                0               0          0   
35        1             1                0               0          0   
36        1             1                0               0          0   

    is_opt_out  is_open  Sent_date       Campaign_type Primary Topic  
0            0        1   2/7/2023  Events - In Person    

In [23]:
email_df[email_df["email_subject"].str.contains("Roadshow")] \
["Primary Topic"].value_counts()

Primary Topic
Ad hoc    1296
/           15
Name: count, dtype: int64

In [24]:
email_df[(email_df["email_subject"].str.contains("Roadshow")) \
& (email_df["Primary Topic"] == "Ad hoc") \
& (email_df["Campaign_type"] == "Event follow up")]

Email_Campaign_id          contact_id  \
42614  70120000000MlGzAAK  00v2000000KLfRoAAL   
42615  70120000000MlGzAAK  00v2000000KLfmGAAT   
42616  70120000000MlGzAAK  00v2000000KMMtsAAH   
42617  70120000000MlGzAAK  00v2000000KLWsiAAH   
42618  70120000000MlGzAAK  00v2000000LnyHcAAJ   
...                   ...                 ...   
43905  70120000000MlGzAAK  00v2000000KNzTiAAL   
43906  70120000000MlGzAAK  00v2000000KNzknAAD   
43907  70120000000MlGzAAK  00v2000000KO0M5AAL   
43908  70120000000MlGzAAK  00v2000000KO0b1AAD   
43909  70120000000MlGzAAK  00v2000000KO0gBAAT   

                         email_subject  is_sent  is_delivered  \
42614  Roadshow Houston (PS) Feedback!        1             1   
42615  Roadshow Houston (PS) Feedback!        1             1   
42616  Roadshow Houston (PS) Feedback!        1             1   
42617  Roadshow Houston (PS) Feedback!        1             1   
42618  Roadshow Houston (PS) Feedback!        1             1   
...                                ...      ...           ...   
43905  Roadshow Houston (PS) Feedback!        1             1   
43906  Roadshow Houston (PS) Feedback!        1             1   
43907  Roadshow Houston (PS) Feedback!        1             1   
43908  Roadshow Houston (PS) Feedback!        1             1   
43909  Roadshow Houston (PS) Feedback!        1             1   

       is_link_clicked  is_hard_bounce  is_bounce  is_opt_out  is_open  \
42614                0               0          0           0        1   
42615                0               0          0           0        0   
42616                0               0          0           0        0   
42617                0               0          0           0        1   
42618                0               0          0           0        0   
...                ...             ...        ...         ...      ...   
43905                0               0          0           0        0   
43906                0               0          0           0        0   
43907                0               0          0           0        0   
43908                0               0          0           0        0   
43909                0               0          0           0        0   

        Sent_date    Campaign_type Primary Topic  
42614  12/28/2023  Event follow up        Ad hoc  
42615  12/28/2023  Event follow up        Ad hoc  
42616  12/28/2023  Event follow up        Ad hoc  
42617  12/28/2023  Event follow up        Ad hoc  
42618  12/28/2023  Event follow up        Ad hoc  
...           ...              ...           ...  
43905  12/28/2023  Event follow up        Ad hoc  
43906  12/28/2023  Event follow up        Ad hoc  
43907  12/28/2023  Event follow up        Ad hoc  
43908  12/28/2023  Event follow up        Ad hoc  
43909  12/28/2023  Event follow up        Ad hoc  

[1296 rows x 13 columns]

In [25]:
email_df[(email_df["email_subject"].str.contains("Roadshow")) \
& (email_df["Primary Topic"] == "Ad hoc") \
& (~email_df["email_subject"].str.contains("Feedback"))]

Empty DataFrame
Columns: [Email_Campaign_id, contact_id, email_subject, is_sent, is_delivered, is_link_clicked, is_hard_bounce, is_bounce, is_opt_out, is_open, Sent_date, Campaign_type, Primary Topic]
Index: []

In [26]:
email_df[(email_df["email_subject"].str.contains("Roadshow")) \
& (email_df["Primary Topic"] != "Ad hoc") \
& (email_df["Primary Topic"] != "/")] \
["Primary Topic"].isnull().value_counts()

Primary Topic
True    5911
Name: count, dtype: int64

/ can be treated as NULL for "Primary Topic"

In [27]:
email_df[email_df["email_subject"].str.contains("Brief")].head()

Email_Campaign_id          contact_id                 email_subject  \
17  701200000006JxpAAE  00v2000000fxdpyAAA  Brief outlook on Hedge Funds   
18  701200000006JxpAAE  00v2000000Mhvw9AAB  Brief outlook on Hedge Funds   
19  701200000006JxpAAE  00v2000000FDzZ6AAL  Brief outlook on Hedge Funds   
20  701200000006JxpAAE  00v2000000f10arAAA  Brief outlook on Hedge Funds   
21  701200000006JxpAAE  00v2000000I2mDSAAZ  Brief outlook on Hedge Funds   

    is_sent  is_delivered  is_link_clicked  is_hard_bounce  is_bounce  \
17        1             1                0               0          0   
18        1             1                0               0          0   
19        1             1                0               0          0   
20        1             1                0               0          0   
21        1             1                0               0          0   

    is_opt_out  is_open  Sent_date       Campaign_type Primary Topic  
17           0        1  8/31/2023  Marketing Campaign         Funds  
18           0        0  8/31/2023  Marketing Campaign         Funds  
19           0        0  8/31/2023  Marketing Campaign         Funds  
20           0        0  8/31/2023  Marketing Campaign         Funds  
21           0        0  8/31/2023  Marketing Campaign         Funds

In [28]:
email_df[email_df["email_subject"].str.contains("Brief outlook")] \
["email_subject"].value_counts()

email_subject
Brief outlook on Portfolio Analytics    495
Brief outlook on Portfolio Insights     382
Brief outlook on Absolute return        375
Brief outlook on Absolute Return        184
Brief outlook on Market Commentary      169
Brief outlook on Liquidity Insights     161
Brief outlook on Defined Benefit         77
Brief outlook on Retirement Link         65
Brief outlook on Investor Funds          44
Brief outlook on Research Quarterly      29
Brief outlook on Healthcare              19
Brief outlook on Equities                14
Brief outlook on Employee                13
Brief outlook on Real Estate             13
Brief outlook on Investment Forum         9
Brief outlook on Hedge Funds              9
Brief outlook on MI-Market Insights       6
Brief outlook on Income                   4
Brief outlook on Variable Annuities       3
Brief outlook on INSTL Global View        3
Brief outlook on Asset Allocation         2
Brief outlook on Global Liquidity         2
Brief outlook on E

In [29]:
email_df[email_df["email_subject"].str.contains("Brief outlook")] \
["Primary Topic"].value_counts()

Primary Topic
Portfolio          877
Market Insights    176
Retirement          65
Funds               53
HealthCare          19
Equities            14
/                   10
Investment          10
Income               4
Asssets              2
Growth               1
Name: count, dtype: int64

In [30]:
email_df[email_df["email_subject"].str.contains("Brief outlook")] \
[["email_subject", "Primary Topic"]].value_counts()

email_subject                         Primary Topic  
Brief outlook on Portfolio Analytics  Portfolio          495
Brief outlook on Portfolio Insights   Portfolio          382
Brief outlook on Market Commentary    Market Insights    169
Brief outlook on Retirement Link      Retirement          65
Brief outlook on Investor Funds       Funds               44
Brief outlook on Healthcare           HealthCare          19
Brief outlook on Equities             Equities            14
Brief outlook on Real Estate          /                   10
Brief outlook on Hedge Funds          Funds                9
Brief outlook on Investment Forum     Investment           9
Brief outlook on MI-Market Insights   Market Insights      6
Brief outlook on Income               Income               4
Brief outlook on Asset Allocation     Asssets              2
Brief outlook on Emerging Markets     Market Insights      1
Brief outlook on Growth               Growth               1
Brief outlook on Investment Tru

In [31]:
email_df[email_df["email_subject"].str.contains("Real Estate")] \
[["email_subject", "Primary Topic"]].value_counts()

email_subject                    Primary Topic
BI Weekly update on Real Estate  Insights         2010
Brief outlook on Real Estate     /                  10
Name: count, dtype: int64

In [32]:
email_df[email_df["email_subject"].str.contains("Brief outlook")] \
["Primary Topic"].isnull().value_counts()

Primary Topic
False    1231
True      851
Name: count, dtype: int64

Breif outlook / cannot identify relevant Primary Topic. 

Can be assigned NULL

In [33]:
email_df[email_df["Primary Topic"] == "--"]

Email_Campaign_id          contact_id         email_subject  is_sent  \
12980  70120000000Lp3eAAC  00v2000000KLpp3AAD  pharma Insights 2023        1   
12981  70120000000Lp3eAAC  00v2000000KKbuRAAT  pharma Insights 2023        1   
12982  70120000000Lp3eAAC  00v2000000KLfRoAAL  pharma Insights 2023        1   
12983  70120000000Lp3eAAC  00v2000000KLfmGAAT  pharma Insights 2023        1   
12984  70120000000Lp3eAAC  00v2000000KMMtsAAH  pharma Insights 2023        1   
12985  70120000000Lp3eAAC  00v2000000KLWsiAAH  pharma Insights 2023        1   
12986  70120000000Lp3eAAC  00v2000000LnyHcAAJ  pharma Insights 2023        1   
12987  70120000000Lp3eAAC  00v2000000Lnxj4AAB  pharma Insights 2023        1   
12988  70120000000Lp3eAAC  00v2000000LnxtKAAR  pharma Insights 2023        1   
12989  70120000000Lp3eAAC  00v2000000LnyKfAAJ  pharma Insights 2023        1   
12990  70120000000Lp3eAAC  00v2000000LnyTqAAJ  pharma Insights 2023        1   
12991  70120000000Lp3eAAC  00v2000000KKdS8AAL  pharma Insights 2023        1   
12992  70120000000Lp3eAAC  00v2000000KKdXVAA1  pharma Insights 2023        1   
12993  70120000000Lp3eAAC  00v2000000KKfmvAAD  pharma Insights 2023        1   
12994  70120000000Lp3eAAC  00v2000000KKjkWAAT  pharma Insights 2023        1   
12995  70120000000Lp3eAAC  00v2000000KKjnYAAT  pharma Insights 2023        1   
12996  70120000000Lp3eAAC  00v2000000KMB2rAAH  pharma Insights 2023        1   
12997  70120000000Lp3eAAC  00v2000000KMB2sAAH  pharma Insights 2023        1   
12998  70120000000Lp3eAAC  00v2000000Lo2lgAAB  pharma Insights 2023        1   
12999  70120000000Lp3eAAC  00v2000000Lo2lhAAB  pharma Insights 2023        1   
13000  70120000000Lp3eAAC  00v2000000Lo3GrAAJ  pharma Insights 2023        1   
13001  70120000000Lp3eAAC  00v2000000Lo51GAAR  pharma Insights 2023        1   
13002  70120000000Lp3eAAC  00v2000000KMUmxAAH  pharma Insights 2023        1   
13003  70120000000Lp3eAAC  00v2000000KMUmyAAH  pharma Insights 2023        1   
13004  70120000000Lp3eAAC  00v2000000KMWU5AAP  pharma Insights 2023        1   
13005  70120000000Lp3eAAC  00v2000000KMWUyAAP  pharma Insights 2023        1   
13006  70120000000Lp3eAAC  00v2000000KMWfDAAX  pharma Insights 2023        1   
13007  70120000000Lp3eAAC  00v2000000Lo2LMAAZ  pharma Insights 2023        1   
13008  70120000000Lp3eAAC  00v2000000Lo2UyAAJ  pharma Insights 2023        1   
13009  70120000000Lp3eAAC  00v2000000Lo2UzAAJ  pharma Insights 2023        1   
13010  70120000000Lp3eAAC  00v2000000KLmA8AAL  pharma Insights 2023        1   
13011  70120000000Lp3eAAC  00v2000000KKjzeAAD  pharma Insights 2023        1   
13012  70120000000Lp3eAAC  00v2000000KKjzfAAD  pharma Insights 2023        1   
13013  70120000000Lp3eAAC  00v2000000KKjzgAAD  pharma Insights 2023        1   
13014  70120000000Lp3eAAC  00v2000000KLt4LAAT  pharma Insights 2023        1   
13015  70120000000Lp3eAAC  00v2000000KLtwDAAT  pharma Insights 2023        1   
13016  70120000000Lp3eAAC  00v2000000KLuNXAA1  pharma Insights 2023        1   
13017  70120000000Lp3eAAC  00v2000000KLuNYAA1  pharma Insights 2023        1   
13018  70120000000Lp3eAAC  00v2000000KLuTNAA1  pharma Insights 2023        1   

       is_delivered  is_link_clicked  is_hard_bounce  is_bounce  is_opt_out  \
12980             1                0               0          0           0   
12981             1                0               0          0           0   
12982             1                0               0          0           0   
12983             1                0               0          0           0   
12984             1                0               0          0           0   
12985             1                0               0          0           0   
12986             1                0               0          0           0   
12987             1                0               0          0           0   
12988             1                0               0          0           0   
12989            

In [34]:
email_df[email_df["email_subject"].str.contains("pharma")] \
["Primary Topic"].value_counts()

Primary Topic
--        39
Ad hoc     9
Bonds      8
Name: count, dtype: int64

In [35]:
email_df[email_df["email_subject"].str.contains("pharma Insights")] \
["Primary Topic"].value_counts()

Primary Topic
--        39
Ad hoc     9
Name: count, dtype: int64

In [36]:
email_df[(email_df["email_subject"].str.contains("pharma Insights")) \
& (email_df["Primary Topic"] == "Ad hoc")]

Email_Campaign_id          contact_id  \
15082  70120000000Lyv6AAC  00v2000000RBiVMAA1   
15083  70120000000Lyv6AAC  00v2000000MOXV1AAP   
15084  70120000000Lyv6AAC  00v2000000MjQC6AAN   
15085  70120000000Lyv6AAC  00v2000000YcB4RAAV   
15086  70120000000Lyv6AAC  00v2000000bFleNAAS   
15087  70120000000Lyv6AAC  00v2000000XPvGnAAL   
15088  70120000000Lyv6AAC  00v2000000YbZLIAA3   
15089  70120000000Lyv6AAC  00v2000000PMD61AAH   
15090  70120000000Lyv6AAC  00v2000000MhGi1AAF   

                                           email_subject  is_sent  \
15082  Your Feedback is important- tell us more on  p...        1   
15083  Your Feedback is important- tell us more on  p...        1   
15084  Your Feedback is important- tell us more on  p...        1   
15085  Your Feedback is important- tell us more on  p...        1   
15086  Your Feedback is important- tell us more on  p...        1   
15087  Your Feedback is important- tell us more on  p...        1   
15088  Your Feedback is important- tell us more on  p...        1   
15089  Your Feedback is important- tell us more on  p...        1   
15090  Your Feedback is important- tell us more on  p...        1   

       is_delivered  is_link_clicked  is_hard_bounce  is_bounce  is_opt_out  \
15082             1                0               0          0           0   
15083             1                0               0          0           0   
15084             1                0               0          0           0   
15085             1                0               0          0           0   
15086             1                0               0          0           0   
15087             1                0               0          0           0   
15088             1                0               0          0           0   
15089             1                0               0          0           0   
15090             1                0               0          0           0   

       is_open  Sent_date    Campaign_type Primary Topic  
15082        1  5/23/2023  Event follow up        Ad hoc  
15083        0  5/23/2023  Event follow up        Ad hoc  
15084        1  5/23/2023  Event follow up        Ad hoc  
15085        0  5/23/2023  Event follow up        Ad hoc  
15086        1  5/23/2023  Event follow up        Ad hoc  
15087        0  5/23/2023  Event follow up        Ad hoc  
15088        0  5/23/2023  Event follow up        Ad hoc  
15089        1  5/23/2023  Event follow up        Ad hoc  
15090        0  5/23/2023  Event follow up        Ad hoc

In [37]:
# email_df[(email_df["email_subject"].str.contains("pharma Insights")) \
# & ~(email_df["Primary Topic"].isnull())]

In [38]:
email_df[email_df["Primary Topic"].isnull()]["email_subject"].value_counts()

email_subject
Roadshow Atlanta (PS)                         2343
Women's Summit Q4                             2296
Roadshow  NYC (PS)                            2067
Alternatives Stratergy trending this year!    1925
Roadshow  Boston (PS)                          972
                                              ... 
Topic of the month Defined Benefit               1
Tax Management fund trending this year!          1
Brief outlook on GL e-Newsletter                 1
Topic of the month Global Liquidity              1
Invstment Acadmey                                1
Name: count, Length: 64, dtype: int64

### Ad hoc vs Adhoc

- Ad hoc and Adhoc are same by looking at the transactions

In [39]:
email_df[(email_df["Primary Topic"] == "Ad hoc")].head()

Email_Campaign_id          contact_id  \
6273  701200000006vcSAAQ  00v2000000KLyuLAAT   
6274  701200000006vcSAAQ  00v2000000KunXJAAZ   
6275  701200000006vcSAAQ  00v2000000KMC1jAAH   
6276  701200000006vcSAAQ  00v2000000KMDCIAA5   
6277  701200000006vcSAAQ  00v2000000KMDCJAA5   

                                         email_subject  is_sent  is_delivered  \
6273  Thanks for attending  Employee Stratergy Session        1             1   
6274  Thanks for attending  Employee Stratergy Session        1             1   
6275  Thanks for attending  Employee Stratergy Session        1             1   
6276  Thanks for attending  Employee Stratergy Session        1             1   
6277  Thanks for attending  Employee Stratergy Session        1             1   

      is_link_clicked  is_hard_bounce  is_bounce  is_opt_out  is_open  \
6273                0               0          0           0        0   
6274                0               0          0           0        1   
6275                0               0          0           0        1   
6276                0               0          0           0        0   
6277                0               0          0           0        0   

     Sent_date    Campaign_type Primary Topic  
6273  7/3/2023  Event follow up        Ad hoc  
6274  7/3/2023  Event follow up        Ad hoc  
6275  7/3/2023  Event follow up        Ad hoc  
6276  7/3/2023  Event follow up        Ad hoc  
6277  7/3/2023  Event follow up        Ad hoc

In [40]:
email_df[(email_df["Primary Topic"] == "Adhoc")].head()

Email_Campaign_id          contact_id  \
13465  70120000000Lp3tAAC  00v2000000KMCeSAAX   
13466  70120000000Lp3tAAC  00v2000000Lny4hAAB   
13467  70120000000Lp3tAAC  00v2000000Lny4iAAB   
13468  70120000000Lp3tAAC  00v2000000Lny2lAAB   
13469  70120000000Lp3tAAC  00v2000000KLuP6AAL   

                                           email_subject  is_sent  \
13465  We are looking forward to seeing you in Birmin...        1   
13466  We are looking forward to seeing you in Birmin...        1   
13467  We are looking forward to seeing you in Birmin...        1   
13468  We are looking forward to seeing you in Birmin...        1   
13469  We are looking forward to seeing you in Birmin...        1   

       is_delivered  is_link_clicked  is_hard_bounce  is_bounce  is_opt_out  \
13465             1                0               0          0           0   
13466             1                0               0          0           0   
13467             1                0               0          0           0   
13468             1                0               0          0           0   
13469             1                0               0          0           0   

       is_open  Sent_date  Campaign_type Primary Topic  
13465        0  8/30/2023  Digital event         Adhoc  
13466        0  8/30/2023  Digital event         Adhoc  
13467        0  8/30/2023  Digital event         Adhoc  
13468        1  8/30/2023  Digital event         Adhoc  
13469        0  8/30/2023  Digital event         Adhoc

In [41]:
email_df[(email_df["Primary Topic"] == "Ad hoc")]["Campaign_type"].value_counts()

Campaign_type
Event follow up    8019
Digital event       613
Name: count, dtype: int64

In [42]:
email_df[(email_df["Primary Topic"] == "Adhoc")]["Campaign_type"].value_counts()

Campaign_type
Digital event    7167
Name: count, dtype: int64

In [43]:
email_df[(email_df["Primary Topic"] == "Ad hoc") \
& (email_df["Campaign_type"] == "Digital event")]["email_subject"].value_counts()

email_subject
You are Invited! Defined Contribution Discussion 2023    613
Name: count, dtype: int64

In [44]:
email_df[(email_df["Primary Topic"] == "Adhoc") \
& (email_df["Campaign_type"] == "Digital event")]["email_subject"].value_counts()

email_subject
Invitation to  Defined Contribution                                              2805
Invitation to  Employee Fund                                                     2606
Meeting for  Product fund Insights                                                781
Invitation to  Global Liquidity                                                   387
We are looking forward to seeing you in Birmingham next week                      286
Meeting for  Fund Insights                                                        108
You are Invited! Global Liquidity Discussion 2023                                  37
Webinar Invite on   Insights Tool                                                  33
Virtual Invitation to attend web conference on Alternatives Stratergy              22
Virtual Invitation to attend web conference on Tax Management fund                 22
Virtual Invitation to attend web conference on Defined Benefit Fund                22
Virtual Invitation to attend web confere

In [45]:
email_df[(email_df["Campaign_type"] == "Event follow up")]["Primary Topic"].value_counts()

Primary Topic
Ad hoc             8019
Portfolio          5307
Market Insights    3210
Equities           1989
ETF                1921
Bonds              1067
Asssets             858
HealthCare          114
Funds                12
Name: count, dtype: int64

### Check flag columns

Journey

- sent
- delivered - bounced
- open
- link clicked / opt out

There are three entries with improper sequence of email opening. Probaly bad data. There were two options: either remove the 3 rows or correct the data. This analysis takes the approach of removing those columns as volume of data is very less.

In [46]:
email_df[(email_df["is_sent"] == 0) \
& (email_df["is_delivered"] == 1)].shape

(0, 13)

In [47]:
email_df[(email_df["is_sent"] == 0) \
& (email_df["is_delivered"] == 0) \
& (email_df["is_open"] == 1)].shape

(0, 13)

In [48]:
email_df[(email_df["is_sent"] == 0) \
& (email_df["is_delivered"] == 0) \
& (email_df["is_bounce"] == 1)].shape

(0, 13)

In [49]:
email_df[["is_bounce", "is_hard_bounce"]].value_counts()

is_bounce  is_hard_bounce
0          0                 125384
1          0                   2764
           1                     88
Name: count, dtype: int64

In [50]:
email_df[["is_open", "is_link_clicked"]].value_counts()

is_open  is_link_clicked
0        0                  94887
1        0                  32757
         1                    589
0        1                      3
Name: count, dtype: int64

In [51]:
email_df[["is_link_clicked", "is_open"]].value_counts()

is_link_clicked  is_open
0                0          94887
                 1          32757
1                1            589
                 0              3
Name: count, dtype: int64

In [52]:
email_df[["is_open", "is_opt_out"]].value_counts()

is_open  is_opt_out
0        0             94890
1        0             33334
         1                12
Name: count, dtype: int64

In [53]:
email_df[["is_open", "is_link_clicked", "is_opt_out"]].value_counts()

is_open  is_link_clicked  is_opt_out
0        0                0             94887
1        0                0             32757
         1                0               577
                          1                12
0        1                0                 3
Name: count, dtype: int64

In [54]:
# email_df[email_df["is_delivered"] == 1][["is_bounce"]]

email_df[["is_delivered", "is_bounce", "is_hard_bounce"]].value_counts().sort_index()

is_delivered  is_bounce  is_hard_bounce
0             1          0                   2702
                         1                     88
1             0          0                 125384
              1          0                     62
Name: count, dtype: int64

### NULL verification

In [55]:
email_df[email_df["Primary Topic"].isnull()]["email_subject"].unique()

array(['Brief outlook on Real Estate', 'pharma Insights 2023',
       'Topic of the month INSTL Global View',
       'Topic of the month Employee', 'pharma Insights Q4',
       'pharma Insights Q1', 'Topic of the month GL e-Newsletter',
       'Topic of the month Liquidity Insights',
       'Topic of the month Global Liquidity',
       'Topic of the month Real Estate',
       'Defined Benefit Fund Update this year!', 'pharma Insights Q2',
       'Topic of the month Absolute return',
       'Topic of the month Absolute Return',
       'Info on  Defined Benefit Stratergy',
       'Defined Contribution Fund Update this year!',
       'Info on  Defined Contribution Stratergy',
       'Info on  Alternatives Fund', 'Brief outlook on Absolute return',
       'Take on  Employee Stratergy', 'Brief outlook on Global Liquidity',
       'Brief outlook on Absolute Return',
       'Brief outlook on GL e-Newsletter',
       'Topic of the month Defined Benefit', 'Roadshow  Boston (PS)',
       'APAC S

In [56]:
email_df[(email_df["Primary Topic"].notnull()) \
& (email_df["email_subject"].str.contains("Topic"))]["Primary Topic"].value_counts()

Primary Topic
Asssets            149
Investment         116
Market Insights     19
Portfolio           19
Funds               17
Equities            11
HealthCare          10
Income               8
Growth               3
Name: count, dtype: int64

In [57]:
email_df[email_df["Primary Topic"] == "Market Insights"]["email_subject"].unique()

array(['Topic of the month Market Commentary',
       'Topic of the month MI-Market Insights',
       'Topic of the month Emerging Markets',
       'Whats Happening around Market Commentary fund',
       'BI Weekly update on Market Commentary',
       'Whats Happening around Long Term Capital Market Assumptions Stratergy',
       'Whats Happening around Market Insights Stratergy',
       'Whats Happening around Market Insights fund',
       'Whats Happening around Market Commentary Stratergy',
       'Whats Happening around Long Term Capital Market Assumptions fund',
       'Brief outlook on MI-Market Insights',
       'BI Weekly update on Emerging Markets',
       'Market Insights fund trending this year!',
       'Long Term Capital Market Assumptions Stratergy trending this year!',
       'Market Commentary Stratergy trending this year!',
       'Market Insights Stratergy trending this year!',
       'Thanks for attending  Market Commentary Stratergy Session',
       'Thanks for atte

### Campaign ID understanding

- Campaign ID is unique for a campaign (set of emails with same subject)

In [58]:
email_df.head()

Email_Campaign_id          contact_id               email_subject  \
0  701200000006cIGAAY  00v2000000IZVYWAA5  Roadshow Houston (Advisor)   
1  701200000006cIGAAY  00v2000000IZW13AAH  Roadshow Houston (Advisor)   
2  701200000006cIGAAY  00v2000000IZUepAAH  Roadshow Houston (Advisor)   
3  701200000006cIGAAY  00v2000000IZWDdAAP  Roadshow Houston (Advisor)   
4  701200000006cIGAAY  00v2000000IZUdSAAX  Roadshow Houston (Advisor)   

   is_sent  is_delivered  is_link_clicked  is_hard_bounce  is_bounce  \
0        1             1                0               0          0   
1        1             1                0               0          0   
2        1             1                0               0          0   
3        1             1                0               0          0   
4        1             1                0               0          0   

   is_opt_out  is_open Sent_date       Campaign_type Primary Topic  
0           0        1  2/7/2023  Events - In Person             /  
1           0        1  2/7/2023  Events - In Person             /  
2           0        0  2/7/2023  Events - In Person             /  
3           0        0  2/7/2023  Events - In Person             /  
4           0        0  2/7/2023  Events - In Person             /

In [59]:
email_df[email_df["Email_Campaign_id"] == "701200000006cIGAAY"]

Email_Campaign_id          contact_id               email_subject  \
0   701200000006cIGAAY  00v2000000IZVYWAA5  Roadshow Houston (Advisor)   
1   701200000006cIGAAY  00v2000000IZW13AAH  Roadshow Houston (Advisor)   
2   701200000006cIGAAY  00v2000000IZUepAAH  Roadshow Houston (Advisor)   
3   701200000006cIGAAY  00v2000000IZWDdAAP  Roadshow Houston (Advisor)   
4   701200000006cIGAAY  00v2000000IZUdSAAX  Roadshow Houston (Advisor)   
5   701200000006cIGAAY  00v2000000IZUkRAAX  Roadshow Houston (Advisor)   
6   701200000006cIGAAY  00v2000000IZVBLAA5  Roadshow Houston (Advisor)   
7   701200000006cIGAAY  00v2000000IZW96AAH  Roadshow Houston (Advisor)   
8   701200000006cIGAAY  00v2000000IZVVVAA5  Roadshow Houston (Advisor)   
9   701200000006cIGAAY  00v2000000IZWDXAA5  Roadshow Houston (Advisor)   
10  701200000006cIGAAY  00v2000000IZVgmAAH  Roadshow Houston (Advisor)   
11  701200000006cIGAAY  00v2000000IZW8gAAH  Roadshow Houston (Advisor)   
12  701200000006cIGAAY  00v2000000IZUjiAAH  Roadshow Houston (Advisor)   
13  701200000006cIGAAY  00v2000000IZW2PAAX  Roadshow Houston (Advisor)   
14  701200000006cIGAAY  00v2000000IZW04AAH  Roadshow Houston (Advisor)   

    is_sent  is_delivered  is_link_clicked  is_hard_bounce  is_bounce  \
0         1             1                0               0          0   
1         1             1                0               0          0   
2         1             1                0               0          0   
3         1             1                0               0          0   
4         1             1                0               0          0   
5         1             1                0               0          0   
6         1             1                0               0          0   
7         1             1                0               0          0   
8         1             1                0               0          0   
9         1             1                0               0          0   
10        1             1                0               0          0   
11        1             1                0               0          0   
12        1             1                0               0          0   
13        1             1                0               0          0   
14        1             1                0               0          0   

    is_opt_out  is_open Sent_date       Campaign_type Primary Topic  
0            0        1  2/7/2023  Events - In Person             /  
1            0        1  2/7/2023  Events - In Person             /  
2            0        0  2/7/2023  Events - In Person             /  
3            0        0  2/7/2023  Events - In Person             /  
4            0        0  2/7/2023  Events - In Person             /  
5            0        0  2/7/2023  Events - In Person             /  
6            0        0  2/7/2023  Events - In Person             /  
7            0        1  2/7/2023  Events - In Person             /  
8            0        0  2/7/2023  Events - In Person             /  
9            0        0  2/7/2023  Events - In Person             /  
10           0        0  2/7/2023  Events - In Person             /  
11           0        1  2/7/2023  Events - In Person             /  
12           0        0  2/7/2023  Events - In Person             /  
13           0        1  2/7/2023  Events - In Person             /  
14           0        0  2/7/2023  Events - In Person             /

### NULL Primary Topic Revisit

- -- and / are part of campaign ids that are NULL

In [60]:
email_df[email_df["Primary Topic"].isnull()].head()

Email_Campaign_id          contact_id                 email_subject  \
37     701200000006JxYAAU  00v2000000MhdkdAAB  Brief outlook on Real Estate   
38     701200000006JxYAAU  00v2000000M5XI2AAN  Brief outlook on Real Estate   
39     701200000006JxYAAU  00v2000000JcxjqAAB  Brief outlook on Real Estate   
12950  70120000000Lp3eAAC  00v2000000KLfjTAAT          pharma Insights 2023   
12951  70120000000Lp3eAAC  00v2000000KLfrfAAD          pharma Insights 2023   

       is_sent  is_delivered  is_link_clicked  is_hard_bounce  is_bounce  \
37           1             1                0               0          0   
38           1             1                0               0          0   
39           1             1                0               0          0   
12950        1             1                0               0          0   
12951        1             1                0               0          0   

       is_opt_out  is_open  Sent_date       Campaign_type Primary Topic  
37              0        0  1/11/2023  Marketing Campaign           NaN  
38              0        0  1/11/2023  Marketing Campaign           NaN  
39              0        1  1/11/2023  Marketing Campaign           NaN  
12950           0        0  10/4/2023  Marketing Campaign           NaN  
12951           0        0  10/4/2023  Marketing Campaign           NaN

In [61]:
email_df[email_df["Email_Campaign_id"] == "701200000006JxYAAU"]

Email_Campaign_id          contact_id                 email_subject  \
27  701200000006JxYAAU  00v2000000LNA9XAAX  Brief outlook on Real Estate   
28  701200000006JxYAAU  00v2000000MhuxqAAB  Brief outlook on Real Estate   
29  701200000006JxYAAU  00v2000000NbHl6AAF  Brief outlook on Real Estate   
30  701200000006JxYAAU  00v2000000P77GRAAZ  Brief outlook on Real Estate   
31  701200000006JxYAAU  00v2000000QGvTdAAL  Brief outlook on Real Estate   
32  701200000006JxYAAU  00v2000000SK6qeAAD  Brief outlook on Real Estate   
33  701200000006JxYAAU  00v2000000fgHbJAAU  Brief outlook on Real Estate   
34  701200000006JxYAAU  00v2000000PL3NlAAL  Brief outlook on Real Estate   
35  701200000006JxYAAU  00v2000000FkmBYAAZ  Brief outlook on Real Estate   
36  701200000006JxYAAU  00v2000000MiYl6AAF  Brief outlook on Real Estate   
37  701200000006JxYAAU  00v2000000MhdkdAAB  Brief outlook on Real Estate   
38  701200000006JxYAAU  00v2000000M5XI2AAN  Brief outlook on Real Estate   
39  701200000006JxYAAU  00v2000000JcxjqAAB  Brief outlook on Real Estate   

    is_sent  is_delivered  is_link_clicked  is_hard_bounce  is_bounce  \
27        1             1                0               0          0   
28        1             1                0               0          0   
29        1             1                0               0          0   
30        1             1                0               0          0   
31        1             1                0               0          0   
32        1             1                0               0          0   
33        1             1                0               0          0   
34        1             1                0               0          0   
35        1             1                0               0          0   
36        1             1                0               0          0   
37        1             1                0               0          0   
38        1             1                0               0          0   
39        1             1                0               0          0   

    is_opt_out  is_open  Sent_date       Campaign_type Primary Topic  
27           0        0  1/11/2023  Marketing Campaign             /  
28           0        0  1/11/2023  Marketing Campaign             /  
29           0        0  1/11/2023  Marketing Campaign             /  
30           0        0  1/11/2023  Marketing Campaign             /  
31           0        1  1/11/2023  Marketing Campaign             /  
32           0        1  1/11/2023  Marketing Campaign             /  
33           0        0  1/11/2023  Marketing Campaign             /  
34           0        0  1/11/2023  Marketing Campaign             /  
35           0        0  1/11/2023  Marketing Campaign             /  
36           0        0  1/11/2023  Marketing Campaign             /  
37           0        0  1/11/2023  Marketing Campaign           NaN  
38           0        0  1/11/2023  Marketing Campaign           NaN  
39           0        1  1/11/2023  Marketing Campaign           NaN

In [62]:
email_df[email_df["Email_Campaign_id"] == "70120000000Lp3eAAC"]["Primary Topic"].value_counts()

Primary Topic
--    39
Name: count, dtype: int64

In [63]:
campaign_ids_null = email_df[email_df["Primary Topic"].isnull()]["Email_Campaign_id"].unique()
email_df[email_df["Email_Campaign_id"].isin(campaign_ids_null)]["Primary Topic"].value_counts()


Primary Topic
--    39
/     10
Name: count, dtype: int64

In [64]:
campaign_ids_null

array(['701200000006JxYAAU', '70120000000Lp3eAAC', '70120000000LrdpAAC',
       '70120000000LrgnAAC', '70120000000LtEzAAK', '70120000000LvNtAAK',
       '70120000000Ly6BAAS', '70120000000Ly6pAAC', '70120000000Ly6PAAS',
       '70120000000Ly6vAAC', '70120000000LyToAAK', '70120000000LyUwAAK',
       '70120000000M1XDAA0', '70120000000M1XjAAK', '70120000000M1Z3AAK',
       '70120000000M1ZIAA0', '70120000000M4r1AAC', '70120000000M62iAAC',
       '70120000000M7HaAAK', '70120000000M8cMAAS', '70120000000MA86AAG',
       '70120000000MAbSAAW', '70120000000MakKAAS', '70120000000MAudAAG',
       '70120000000MCmOAAW', '70120000000MCXJAA4', '70120000000MdBcAAK',
       '70120000000MdDEAA0', '70120000000MG6SAAW', '70120000000MgQyAAK',
       '70120000000MhdCAAS', '70120000000Mhi5AAC', '70120000000MhwkAAC',
       '70120000000Mi3RAAS', '70120000000MiOmAAK', '70120000000MjuxAAC',
       '70120000000MLUvAAO', '70120000000MNIiAAO', '70120000000MNJIAA4',
       '70120000000MNKBAA4', '70120000000MNKGAA4', 